turbopack/crates/turbo-tasks-testing/tests/performance.rs

```diff
@@ -0,0 +1,100 @@
+#![feature(arbitrary_self_types)]
+
+use std::time::Duration;
+
+use turbo_tasks::Vc;
+use turbo_tasks_testing::{register, run, Registration};
+
+static REGISTRATION: Registration = register!();
+
+const COUNT1: u32 = 100;
+const COUNT2: u32 = 2000;
+
+#[tokio::test]
+async fn many_calls_to_many_children() {
+    run(&REGISTRATION, || async {
+        // The first call will actually execute many_children and its children.
+        let start = std::time::Instant::now();
+        calls_many_children(0).strongly_consistent().await?;
+        println!("Initial call took {:?}", start.elapsed());
+
+        // The second call will connect to the cached many_children, but it would be ok if that's
+        // not yet optimized.
+        let start = std::time::Instant::now();
+        calls_many_children(1).strongly_consistent().await?;
+        println!("Second call took {:?}", start.elapsed());
+
+        // Susbsequent calls should be very fast.
+        let start = std::time::Instant::now();
+        for i in 2..COUNT1 {
+            calls_many_children(i).strongly_consistent().await?;
+        }
+        let subsequent = start.elapsed();
+        println!(
+            "First {} subsequent calls took {:?}",
+            COUNT1 - 2,
+            subsequent
+        );
+
+        let start = std::time::Instant::now();
+        for i in COUNT1..COUNT1 * 2 - 2 {
+            calls_many_children(i).strongly_consistent().await?;
+        }
+        let subsequent2 = start.elapsed();
+        println!(
+            "Another {} subsequent calls took {:?}",
+            COUNT1 - 2,
+            subsequent2
+        );
+
+        let start = std::time::Instant::now();
+        calls_many_children(COUNT1 - 1)
+            .strongly_consistent()
+            .await?;
+        let final_call = start.elapsed();
+        println!("Final call took {:?}", final_call);
+
+        assert!(
+            subsequent2 * 2 < subsequent * 3,
+            "Performance should not regress with more calls"
+        );
+
+        assert!(
+            subsequent < Duration::from_micros(100) * (COUNT1 - 2),
+            "Each call should be less than 100µs"
+        );
+
+        assert!(
+            subsequent2 < Duration::from_micros(100) * (COUNT1 - 2),
+            "Each call should be less than 100µs"
+        );
```

Comment:
I love benchmarks, but I don't like performance-related tests like this because they could be flaky, even if the threshold seems generous:

- It's quite common that the CI server is overloaded by trying to run many tests in parallel, and it's quite possible the test could be heavily resource starved (e.g. the machine is swapping) based on whatever other random collection of tests are running at the moment.
- Tests run on `debug` binaries by default, so this probably isn't representative of real performance anyways.

Make it a benchmark, and I think we can commit to getting codspeed in place sooner rather than later.

Reply:
I'm already seeing this test flake on real PRs: https://github.com/vercel/next.js/actions/runs/11240310104/job/31249465632

![Screenshot 2024-10-08 at 11-01-20 hand-fix some code gated behind next-swc-napi_plugin · vercel_next js@7163d5b](https://github.com/user-attachments/assets/d36cd6b3-8f64-4631-8e2a-41f6f6557d83)